code/mission/missionparse.cpp

```diff
@@ -5477,41 +5477,55 @@ void parse_variables()
 		return;
 	}
 
-	// Goober5000 - now set the default value, if it's a campaign-persistent variable
+	// Goober5000 - now set the default value, if it's a variable saved on mission progress
 	// loop through the current mission's variables
 	for (j = 0; j < num_variables; j++) {
 		// check against existing variables
-		for (i = 0; i < Campaign.num_variables; i++) {
-			// if the active mission has a variable with the same name as a campaign
-			// variable AND it is not a block variable, override its initial value
-			// with the previous mission's value
-			if ( !stricmp(Sexp_variables[j].variable_name, Campaign.variables[i].variable_name) ) {
-				if (Sexp_variables[j].type  & SEXP_VARIABLE_CAMPAIGN_PERSISTENT) {
-					Sexp_variables[j].type = Campaign.variables[i].type;
-					strcpy_s(Sexp_variables[j].text, Campaign.variables[i].text);
+		for (i = 0; i < Campaign.persistent_variables.size(); i++) {
+			// if the active mission has a variable with the same name as a variable saved to the campaign file override its initial value with the previous mission's value
+			if ( !stricmp(Sexp_variables[j].variable_name, Campaign.persistent_variables[i].variable_name) ) {
+				// if this is an eternal that shares the same name as a non-eternal warn but do nothing
+				if (Sexp_variables[j].type & SEXP_VARIABLE_SAVE_TO_PLAYER_FILE) {
+					Warning(LOCATION, "Variable %s is marked eternal but has the same name as another persistent variable. One of these should be renamed to avoid confusion", Sexp_variables[j].text);
```

Comment:
Since this is inside a parse function you should use `error_display` here to make it easier to find the location in the mission file in case a manual edit is needed.